File: tfg-antigravity-docs/src/data/translate.py
import re

translations = {
  "Demanda Peninsular": "Procura Peninsular",
  "Total Load — ES + PT": "Carga Total — ES + PT",
  "Programa de Producción": "Programa de Produção",
  "Potencia Disponible por Tecnología": "Potência Disponível por Tecnologia",
  "Installed Capacity per Type": "Capacidade Instalada por Tipo",
  "Actual Generation per Unit": "Geração Real por Unidade",
  "Water Reservoirs & Hydro": "Reservatórios de Água e Hidroelétrica",
  "Otros Indicadores — CO₂ + Renovables": "Outros Indicadores — CO₂ + Renováveis",
  "Generación libre de CO₂ (%)": "Geração livre de CO₂ (%)",
  "Mercados y Precios — SPOT vs PVPC": "Mercados e Preços — SPOT vs PVPC",
  "Precio Final Desglosado": "Preço Final Detalhado",
  "Energy Prices — Europa": "Preços de Energia — Europa",
  "Precios de Desvíos Tiempo Real": "Preços de Desvios em Tempo Real",
  "Precios de Desvíos Tiempo Real (28-29 Abril)": "Preços de Desvios em Tempo Real (28-29 Abril)",
  "Saldos por Frontera (P48)": "Saldos por Fronteira (P48)",
  "Saldos Horarios por Frontera (28-29 Abril)": "Saldos Horários por Fronteira (28-29 Abril)",
  "Cross-Border Physical Flows": "Fluxos Físicos Transfronteiriços",
  "Scheduled Commercial Exchanges": "Intercâmbios Comerciais Programados",
  "Subastas Explícitas de Capacidad": "Leilões Explícitos de Capacidade",
  "Forecast Transfer Capacities": "Capacidades de Transferência Previstas",
  "Current Balancing State": "Estado Atual de Balanceamento",
  "Imbalance — Volumen MW": "Desequilíbrio — Volume MW",
  "Imbalance Prices": "Preços de Desequilíbrio",
  "FRR Capacity — Reservas": "Capacidade FRR — Reservas",
  "Cost of Congestion Management": "Custo de Gestão de Congestionamento",
  "Fall-backs y Protocolos Emergencia": "Fall-backs e Protocolos de Emergência",
  "Evolución de la demanda eléctrica en España — 28 y 29 de abril de 2025": "Evolução da procura de eletricidade em Espanha — 28 e 29 de abril de 2025",
  "Demanda real frente a previsión day-ahead — España y Portugal": "Procura real face à previsão day-ahead — Espanha e Portugal",
  "Desglose de la energía programada en los mercados de producción": "Repartição da energia programada nos mercados de produção",
  "Comparativa de potencia disponible por tecnología": "Comparação de potência disponível por tecnologia",
  "Capacidad instalada del sistema eléctrico peninsular por tipo de tecnología": "Capacidade instalada do sistema elétrico peninsular por tipo de tecnologia",
  "Generación real por unidad de producción — Estado del mix durante el colapso": "Geração real por unidade de produção — Estado do mix durante o colapso",
  "Almacenamiento en embalses y plantas hidroeléctricas — España, 2025": "Armazenamento em albufeiras e centrais hidroelétricas — Espanha, 2025",
  "Porcentaje de generación libre de CO₂ y previsiones renovables": "Percentagem de geração livre de CO₂ e previsões renováveis",
  "Comparativa de precios eléctricos — Mercado SPOT mayorista vs. PVPC 2.0TD": "Comparação de preços de eletricidade — Mercado grossista SPOT vs. PVPC 2.0TD",
  "Componentes del precio final de la electricidad": "Componentes do preço final da eletricidade",
  "Precios day-ahead del mercado mayorista ibérico (OMIE)": "Preços day-ahead do mercado grossista ibérico (OMIE)",
  "Precios de desvíos y restricciones técnicas en tiempo real": "Preços de desvios e restrições técnicas em tempo real",
  "Saldos netos de intercambio programados (P48)": "Saldos líquidos de intercâmbio programados (P48)",
  "Flujos físicos reales en las interconexiones de España": "Fluxos físicos reais nas interconexões de Espanha",
  "Saldo neto de intercambios comerciales programados": "Saldo líquido de intercâmbios comerciais programados",
  "Subastas explícitas de capacidad de interconexión": "Leilões explícitos de capacidade de interconexão",
  "Capacidades de transferencia previstas en las interconexiones": "Capacidades de transferência previstas nas interconexões",
  "Evolución del desequilibrio eléctrico (imbalance) — España": "Evolução do desequilíbrio elétrico (imbalance) — Espanha",
  "Volumen de desequilibrio entre generación y demanda — España": "Volume de desequilíbrio entre geração e procura — Espanha",
  "Precios de los desvíos de balance — España": "Preços dos desvios de balanceamento — Espanha",
  "Capacidad real y prevista de Reserva de Restauración de Frecuencia": "Capacidade real e prevista de Reserva de Restauração de Frequência",
  "Costes mensuales de gestión de congestión (countertrading)": "Custos mensais de gestão de congestionamento (countertrading)",
  "Desconexiones de TSOs y activación de procedimientos de contingencia": "Desconexões de TSOs e ativação de procedimentos de contingência",
  "Demanda Real vs. Programada vs. Prevista · resolución 5 minutos": "Procura Real vs. Programada vs. Prevista · resolução de 5 minutos",
  "CTA|ES y CTA|PT · 28–29 de abril de 2025 · MTU horario": "CTA|ES e CTA|PT · 28–29 de abril de 2025 · MTU horário",
  "España, 28 de abril de 2025 (MWh)": "Espanha, 28 de abril de 2025 (MWh)",
  "Madrugada del 28 y 29 de abril de 2025 (MW)": "Madrugada de 28 e 29 de abril de 2025 (MW)",
  "2025 (MW) — Iberian Peninsula": "2025 (MW) — Península Ibérica",
  "28 de abril de 2025 (MW)": "28 de abril de 2025 (MW)",
  "MWh por semana": "MWh por semana",
  "27–29 de abril de 2025": "27–29 de abril de 2025",
  "28–29 de abril de 2025, €/MWh": "28–29 de abril de 2025, €/MWh",
  "España, 28 de abril de 2025, hora 00:00 CEST (€/MWh)": "Espanha, 28 de abril de 2025, hora 00:00 CEST (€/MWh)",
  "España y Portugal, 28–29 de abril de 2025 (€/MWh)": "Espanha e Portugal, 28–29 de abril de 2025 (€/MWh)",
  "28–29 de abril de 2025 (€/MWh)": "28–29 de abril de 2025 (€/MWh)",
  "España con Francia, Portugal, Marruecos y Andorra · 28–29 de abril de 2025 (MW/h)": "Espanha com França, Portugal, Marrocos e Andorra · 28–29 de abril de 2025 (MW/h)",
  "España–Francia y España–Portugal · 28–29 de abril de 2025 (MW)": "Espanha–França e Espanha–Portugal · 28–29 de abril de 2025 (MW)",
  "España con Francia y Portugal · 28 de abril de 2025 (MW, resolución 15 min)": "Espanha com França e Portugal · 28 de abril de 2025 (MW, resolução 15 min)",
  "España–Francia y España–Portugal, abril 2025 (MW y €/MWh)": "Espanha–França e Espanha–Portugal, abril de 2025 (MW e €/MWh)",
  "28–29 de abril de 2025 (MWh cada 15 minutos)": "28–29 de abril de 2025 (MWh a cada 15 minutos)",
  "28–29 de abril de 2025 (MWh cada 15 min)": "28–29 de abril de 2025 (MWh a cada 15 min)",
  "España — 2025 (MW por trimestre)": "Espanha — 2025 (MW por trimestre)",
  "España, enero–diciembre 2025 (€)": "Espanha, janeiro–dezembro de 2025 (€)",
  "Europa, 28–29 de abril de 2025": "Europa, 28–29 de abril de 2025",
  "Demanda y Sistema": "Procura e Sistema",
  "Generación y Capacidad": "Geração e Capacidade",
  "Mercados y Precios": "Mercados e Preços",
  "Interconexiones y Flujos": "Interconexões e Fluxos",
  "Balance, Estabilidad y Respuesta": "Balanço, Estabilidade e Resposta",
  "¿Qué se esperaba que consumiera el sistema y qué ocurrió realmente?": "O que se esperava que o sistema consumisse e o que aconteceu realmente?",
  "¿Qué podía generar el sistema, qué estaba generando y por qué tecnología?": "O que podia o sistema gerar, o que estava a gerar e por qual tecnologia?",
  "¿Qué señales económicas emitió el sistema antes, durante y después del colapso?": "Que sinais económicos emitiu o sistema antes, durante e depois do colapso?",
  "¿Cómo se comportaron las fronteras eléctricas antes, durante y después del apagón?": "Como se comportaram as fronteiras elétricas antes, durante e depois do apagão?",
  "¿Cómo perdió el operador el control del sistema y cómo intentó recuperarlo?": "Como o operador perdeu o controlo do sistema e como tentou recuperá-lo?",
  "Primera oscilación": "Primeira oscilação",
  "Segunda oscilación": "Segunda oscilação",
  "Pérdida de sincronismo": "Perda de sincronismo",
  "Islandización ibérica": "Isolamento ibérico",
  "Pico de déficit": "Pico de défice",
  "Inicio recuperación": "Início da recuperação",
  "Restauración completa": "Restauração completa"
}

def translate_file():
    with open('forensicCharts.js', 'r', encoding='utf-8') as f:
        text = f.read()

    # Apply translations for single string values
    for k, v in translations.items():
        # Look for the exact string surrounded by single quotes
        pattern = r"'" + re.escape(k) + r"'"
        replacement = r"'" + v + r"'"
        text = re.sub(pattern, replacement, text)

    # Process chart blocks
    # We want to replace desc: ... and rel: ... with desc_pt: ... and rel_pt: ...
    # And delete the other language variants.
    
    # We can use a regex to capture each chart block and modify it
    def process_chart(match):
        chart_text = match.group(0)
        
        # extract desc_pt and rel_pt
        desc_pt_match = re.search(r"desc_pt:\s*`([\s\S]*?)`,", chart_text)
        rel_pt_match = re.search(r"rel_pt:\s*`([\s\S]*?)`,", chart_text)
        
        if desc_pt_match and rel_pt_match:
            desc_pt_content = desc_pt_match.group(1)
            rel_pt_content = rel_pt_match.group(1)
            
            # replace desc and rel
            chart_text = re.sub(r"desc:\s*`[\s\S]*?`,", "desc: `" + desc_pt_content + "`,", chart_text, count=1)
            chart_text = re.sub(r"rel:\s*`[\s\S]*?`,", "rel: `" + rel_pt_content + "`,", chart_text, count=1)
            
            # delete desc_en, rel_en, desc_pt, rel_pt, desc_fr, rel_fr, desc_it, rel_it, desc_de, rel_de
            languages = ['en', 'pt', 'fr', 'it', 'de']
            for lang in languages:
                chart_text = re.sub(r"desc_" + lang + r":\s*`[\s\S]*?`,\n?", "", chart_text)
                chart_text = re.sub(r"rel_" + lang + r":\s*`[\s\S]*?`,\n?", "", chart_text)
                
            return chart_text
        return chart_text

    text = re.sub(r"\{\s*id:\s*'chart-\d+'[\s\S]*?^\s*\},?", process_chart, text, flags=re.MULTILINE)

    with open('forensicCharts_pt.js', 'w', encoding='utf-8') as f:
        f.write(text)

if __name__ == '__main__':
    translate_file()

File: tfg-antigravity-docs/src/data/translate.js
const fs = require('fs');

try {
  let content = fs.readFileSync('C:\\Users\\aphmo\\Proyectos\\TFG OVERLEAF\\tfg-antigravity-docs\\src\\data\\forensicCharts.js', 'utf8');

  // 1. Replace categories and timeline
  const simpleTranslations = {
    // CATEGORIES
    "'Demanda y Sistema'": "'Nachfrage und System'",
    "'¿Qué se esperaba que consumiera el sistema y qué ocurrió realmente?'": "'Was wurde vom System erwartet und was geschah wirklich?'",
    "'Generación y Capacidad'": "'Erzeugung und Kapazität'",
    "'¿Qué podía generar el sistema, qué estaba generando y por qué tecnología?'": "'Was konnte das System erzeugen, was erzeugte es und mit welcher Technologie?'",
    "'Mercados y Precios'": "'Märkte und Preise'",
    "'¿Qué señales económicas emitió el sistema antes, durante y después del colapso?'": "'Welche wirtschaftlichen Signale sendete das System vor, während und nach dem Zusammenbruch?'",
    "'Interconexiones y Flujos'": "'Verbindungen und Flüsse'",
    "'¿Cómo se comportaron las fronteras eléctricas antes, durante y después del apagón?'": "'Wie verhielten sich die Stromgrenzen vor, während und nach dem Stromausfall?'",
    "'Balance, Estabilidad y Respuesta'": "'Gleichgewicht, Stabilität und Reaktion'",
    "'¿Cómo perdió el operador el control del sistema y cómo intentó recuperarlo?'": "'Wie verlor der Betreiber die Kontrolle über das System und wie versuchte er, sie wiederzuerlangen?'",

    // TIMELINE_MARKERS
    "'Primera oscilación'": "'Erste Schwingung'",
    "'Segunda oscilación'": "'Zweite Schwingung'",
    "'Pérdida de sincronismo'": "'Verlust der Synchronität'",
    "'Islandización ibérica'": "'Iberische Inselbildung'",
    "'Pico de déficit'": "'Spitze des Defizits'",
    "'Inicio recuperación'": "'Beginn der Erholung'",
    "'Restauración completa'": "'Vollständige Wiederherstellung'",

    // Titles and strings
    "'Demanda Peninsular'": "'Nachfrage der Halbinsel'",
    "'Evolución de la demanda eléctrica en España — 28 y 29 de abril de 2025'": "'Entwicklung der Stromnachfrage in Spanien — 28. und 29. April 2025'",
    "'Demanda Real vs. Programada vs. Prevista · resolución 5 minutos'": "'Reale Nachfrage vs. Geplant vs. Prognostiziert · 5-Minuten-Auflösung'",

    "'Total Load — ES + PT'": "'Gesamtlast — ES + PT'",
    "'Demanda real frente a previsión day-ahead — España y Portugal'": "'Reale Nachfrage im Vergleich zur Day-Ahead-Prognose — Spanien und Portugal'",
    "'CTA|ES y CTA|PT · 28–29 de abril de 2025 · MTU horario'": "'CTA|ES und CTA|PT · 28.–29. April 2025 · Stündliche MTU'",

    "'Programa de Producción'": "'Produktionsprogramm'",
    "'Desglose de la energía programada en los mercados de producción'": "'Aufschlüsselung der in den Produktionsmärkten geplanten Energie'",
    "'España, 28 de abril de 2025 (MWh)'": "'Spanien, 28. April 2025 (MWh)'",

    "'Potencia Disponible por Tecnología'": "'Verfügbare Leistung nach Technologie'",
    "'Comparativa de potencia disponible por tecnología'": "'Vergleich der verfügbaren Leistung nach Technologie'",
    "'Madrugada del 28 y 29 de abril de 2025 (MW)'": "'Frühe Morgenstunden 28. und 29. April 2025 (MW)'",

    "'Installed Capacity per Type'": "'Installierte Kapazität nach Typ'",
    "'Capacidad instalada del sistema eléctrico peninsular por tipo de tecnología'": "'Installierte Kapazität des iberischen Stromsystems nach Technologie'",
    "'2025 (MW) — Iberian Peninsula'": "'2025 (MW) — Iberische Halbinsel'",

    "'Generación Real por Tecnología'": "'Reale Erzeugung nach Technologie'",
    "'Generación eléctrica en España por tecnología'": "'Stromerzeugung in Spanien nach Technologie'",
    "'Días 28 y 29 de abril de 2025 — Península (MW)'": "'28. und 29. April 2025 — Halbinsel (MW)'",

    "'Precios SPOT y Desacoplamiento'": "'SPOT-Preise und Entkopplung'",
    "'Evolución del precio en el mercado diario (Day-Ahead)'": "'Preis-Entwicklung im Day-Ahead-Markt'",
    "'España (MIBEL) frente a Francia y Alemania — Abril 2025 (€/MWh)'": "'Spanien (MIBEL) im Vergleich zu Frankreich und Deutschland — April 2025 (€/MWh)'",

    "'Precios Intradiarios y Restricciones'": "'Intraday-Preise und Einschränkungen'",
    "'Evolución de los mercados intradiarios y servicios de ajuste'": "'Entwicklung der Intraday-Märkte und Anpassungsdienste'",
    "'España, 28 de abril de 2025 (€/MWh)'": "'Spanien, 28. April 2025 (€/MWh)'",

    "'Coste de Restricciones Fase I'": "'Kosten der Phase-I-Einschränkungen'",
    "'Coste horario del proceso de solución de restricciones técnicas (PBF)'": "'Stündliche Kosten des technischen Einschränkungslösungsprozesses (PBF)'",

    "'Energía y Coste RR/mFRR'": "'Energie und Kosten RR/mFRR'",
    "'Energía asignada y precio de las reservas de sustitución y terciaria'": "'Zugelassene Energie und Preis der Ersatz- und Tertiärreserven'",
    "'España — Abril 2025 (MW y €/MWh)'": "'Spanien — April 2025 (MW und €/MWh)'",

    "'Margen de Reserva y Precios'": "'Reserve-Marge und Preise'",
    "'Correlación entre el margen de reserva operativa y los precios de mercado'": "'Korrelation zwischen operativer Reserve-Marge und Marktpreisen'",
    "'Península Ibérica — 2025'": "'Iberische Halbinsel — 2025'",

    "'Rentas de Congestión PBF'": "'PBF-Engpasserlöse'",
    "'Evolución de las rentas de congestión en el proceso básico de facturación'": "'Entwicklung der Engpasserlöse im grundlegenden Abrechnungsprozess'",

    "'Saldos por Frontera (P48)'": "'Salden nach Grenze (P48)'",
    "'Saldos netos de intercambio programados (P48)'": "'Geplante Netto-Austauschsalden (P48)'",
    "'España con Francia, Portugal, Marruecos y Andorra — 28-29 de abril de 2025 (MW/h)'": "'Spanien mit Frankreich, Portugal, Marokko und Andorra — 28.-29. April 2025 (MW/h)'",

    "'Cross-Border Physical Flows'": "'Grenzüberschreitende physische Flüsse'",
    "'Flujos físicos reales en las interconexiones de España'": "'Tatsächliche physische Flüsse auf den spanischen Verbindungsleitungen'",
    "'España-Francia y España-Portugal — 28-29 de abril de 2025 (MW)'": "'Spanien-Frankreich und Spanien-Portugal — 28.-29. April 2025 (MW)'",

    "'Scheduled Commercial Exchanges'": "'Geplante kommerzielle Austausche'",
    "'Saldo neto de intercambios comerciales programados'": "'Netto-Saldo der geplanten kommerziellen Austausche'",
    "'España con Francia y Portugal — 28 de abril de 2025 (MW, resolución 15 min)'": "'Spanien mit Frankreich und Portugal — 28. April 2025 (MW, 15-Min.-Auflösung)'",

    "'Subastas Explícitas de Capacidad'": "'Explizite Kapazitätsauktionen'",
    "'Subastas explícitas de capacidad de interconexión'": "'Explizite Auktionen der Verbindungskapazität'",
    "'España-Francia y España-Portugal, abril 2025 (MW y €/MWh)'": "'Spanien-Frankreich und Spanien-Portugal, April 2025 (MW und €/MWh)'",

    "'Forecast Transfer Capacities'": "'Prognostizierte Transferkapazitäten'",
    "'Capacidades de transferencia previstas en las interconexiones'": "'Geplante Transferkapazitäten der Verbindungsleitungen'",

    "'Current Balancing State'": "'Aktueller Ausgleichszustand'",
    "'Evolución del desequilibrio eléctrico (imbalance) - España'": "'Entwicklung des elektrischen Ungleichgewichts (Imbalance) - Spanien'",
    "'28-29 de abril de 2025 (MWh cada 15 minutos)'": "'28.-29. April 2025 (MWh alle 15 Minuten)'",

    "'Imbalance - Volumen MW'": "'Imbalance - Volumen MW'",
    "'Volumen de desequilibrio entre generación y demanda - España'": "'Volumen des Ungleichgewichts zwischen Erzeugung und Nachfrage - Spanien'",
    "'28-29 de abril de 2025 (MWh cada 15 min)'": "'28.-29. April 2025 (MWh alle 15 Min)'",

    "'Imbalance Prices'": "'Ausgleichspreise'",
    "'Precios de los desvíos de balance - España'": "'Preise für Bilanzabweichungen - Spanien'",
    "'28-29 de abril de 2025 (€/MWh)'": "'28.-29. April 2025 (€/MWh)'",

    "'FRR Capacity — Reservas'": "'FRR-Kapazität — Reserven'",
    "'Capacidad real y prevista de Reserva de Restauración de Frecuencia'": "'Tatsächliche und geplante Frequency Restoration Reserve (FRR)-Kapazität'",
    "'España — 2025 (MW por trimestre)'": "'Spanien — 2025 (MW pro Quartal)'",

    "'Cost of Congestion Management'": "'Kosten des Engpassmanagements'",
    "'Costes mensuales de gestión de congestión (countertrading)'": "'Monatliche Kosten für das Engpassmanagement (Countertrading)'",
    "'España, enero–diciembre 2025 (€)'": "'Spanien, Januar–Dezember 2025 (€)'",

    "'Fall-backs y Protocolos Emergencia'": "'Fall-backs und Notfallprotokolle'",
    "'Desconexiones de TSOs y activación de procedimientos de contingencia'": "'Abschaltungen von TSOs und Aktivierung von Notfallverfahren'",
    "'Europa, 28–29 de abril de 2025'": "'Europa, 28.–29. April 2025'"
  };

  for (const [es, de] of Object.entries(simpleTranslations)) {
    content = content.split(es).join(de);
  }

  // Handle some differences in quotes / spacing just in case
  const moreTranslations = {
    "'España con Francia y Portugal — 28 de abril de 2025 (MW, resolución 15 min)'": "'Spanien mit Frankreich und Portugal — 28. April 2025 (MW, 15-Min.-Auflösung)'",
  };
  for (const [es, de] of Object.entries(moreTranslations)) {
    content = content.split(es).join(de);
  }

  // 2. Replace desc and rel with desc_de and rel_de
  const chartsSplit = content.split('id: \'chart-');
  for (let i = 1; i < chartsSplit.length; i++) {
    let chunk = chartsSplit[i];
    
    // Extract desc_de
    const descDeMatch = chunk.match(/desc_de:\s*`([\s\S]*?)`/);
    if (descDeMatch) {
      const descDeText = descDeMatch[1];
      chunk = chunk.replace(/desc:\s*`[\s\S]*?`,\s*rel:/, `desc: \`${descDeText}\`,\n    rel:`);
    }
    
    // Extract rel_de
    const relDeMatch = chunk.match(/rel_de:\s*`([\s\S]*?)`/);
    if (relDeMatch) {
      const relDeText = relDeMatch[1];
      chunk = chunk.replace(/rel:\s*`[\s\S]*?`,\s*desc_en:/, `rel: \`${relDeText}\`,\n    desc_en:`);
    }
    
    chartsSplit[i] = chunk;
  }

  content = chartsSplit.join('id: \'chart-');

  fs.writeFileSync('C:\\Users\\aphmo\\Proyectos\\TFG OVERLEAF\\tfg-antigravity-docs\\src\\data\\forensicCharts_de.js', content, 'utf8');
  console.log('Successfully wrote forensicCharts_de.js');
} catch (error) {
  console.error('Error:', error);
}
